src/snowflake/snowpark/_internal/analyzer/metadata_utils.py

```diff
@@ -182,3 +188,27 @@ def infer_metadata(
             quoted_identifiers = None
 
     return PlanMetadata(attributes=attributes, quoted_identifiers=quoted_identifiers)
+
+
+def cache_metadata_on_select_statement(
+    source_plan: Optional[LogicalPlan], metadata: PlanMetadata
```

Comment:
since this function is called  cache_metadata_on_select_statement, let's make the caller do a check and make sure it passes a SelectStatement node inside. 
Or call this function cache_metadata_if_select_statement

Reply:
I'd prefer to call this function cache_metadata_if_select_statement and keep all logic here (more clear)